File: src/main/java/com/nobility/downloader/downloads/Download.java
package com.nobility.downloader.downloads;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.nobility.downloader.scraper.Episode;
import com.nobility.downloader.utils.StringChecker;
import com.nobility.downloader.utils.Tools;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;

import java.io.File;

public class Download {

    private String downloadPath;
    private final String name;
    private final String dateAdded;
    private final String url;
    private long fileSize;
    private String seriesLink;

    //used for the session, doesn't need to be saved.
    @JsonIgnore
    private boolean downloading = false;
    @JsonIgnore
    private boolean queued = false;
    @JsonIgnore
    private final StringProperty progress;

    public Download(String downloadPath, String name, String dateAdded, String url) {
        this.downloadPath = downloadPath;
        this.name = name;
        this.dateAdded = dateAdded;
        this.url = url;
        progress = new SimpleStringProperty("0%");
    }

    public void updateProgress() {
        if (queued) {
            setProgressValue("Queued");
            return;
        }
        File video = new File(downloadPath);
        if (video.exists()) {
            double ratio = video.length() / (double) fileSize;
            setProgressValue(Tools.percentFormat.format(ratio));
        } else {
            setProgressValue("File Not Found");
        }
    }

    private void setProgressValue(String value) {
        if (!progress.getValue().equals(value)) {
            progress.setValue(value);
        }
    }

    public void update(Download download) {
        fileSize = download.fileSize;
        downloading = download.downloading;
        updateProgress();
    }

    public String getName() {
        return name;
    }

    public String getDownloadPath() {
        return downloadPath;
    }

    public void setDownloadPath(String downloadPath) {
        if (!this.downloadPath.equals(downloadPath)) {
            this.downloadPath = downloadPath;
        }
    }

    @JsonIgnore
    public File getDownloadFile() {
        if (!StringChecker.isNullOrEmpty(downloadPath)) {
            File file = new File(downloadPath);
            if (file.exists()) {
                return file;
            }
        }
        return null;
    }

    @JsonIgnore
    public boolean isComplete() {
        if (fileSize <= 0) {
            return false;
        }
        File file = getDownloadFile();
        if (file != null) {
            if (file.length() <= 0) {
                return false;
            }
            return file.length() >= fileSize;
        }
        return false;
    }

    public String getUrl() {
        return url;
    }

    public long getFileSize() {
        return fileSize;
    }

    public void setFileSize(long fileSize) {
        this.fileSize = fileSize;
    }

    public boolean isDownloading() {
        return downloading;
    }

    public void setDownloading(boolean downloading) {
        this.downloading = downloading;
    }

    public void setSeriesLink(String seriesLink) {
        this.seriesLink = seriesLink;
    }

    public String getSeriesLink() {
        return seriesLink;
    }

    public String getDateAdded() {
        return dateAdded;
    }

    public StringProperty getProgress() {
        return progress;
    }

    public boolean isQueued() {
        return queued;
    }

    public void setQueued(boolean queued) {
        this.queued = queued;
    }

    public Episode toEpisode() {
        return new Episode(name, url, seriesLink);
    }
}

File: src/main/java/com/nobility/downloader/downloads/Download.kt
package com.nobility.downloader.downloads

import com.fasterxml.jackson.annotation.JsonIgnore
import com.nobility.downloader.cache.Episode
import com.nobility.downloader.utils.StringChecker
import com.nobility.downloader.utils.Tools
import javafx.beans.property.SimpleStringProperty
import javafx.beans.property.StringProperty
import java.io.File

class Download(var downloadPath: String, val dateAdded: String, val episode: Episode) {
    var fileSize: Long = 0

    //used for the session, doesn't need to be saved.
    @JsonIgnore
    var isDownloading = false

    @JsonIgnore
    var isQueued = false

    @get:JsonIgnore
    @JsonIgnore
    val progress: StringProperty

    init {
        progress = SimpleStringProperty("0%")
    }

    fun updateProgress() {
        if (isQueued) {
            setProgressValue("Queued")
            return
        }
        val video = File(downloadPath)
        if (video.exists()) {
            val ratio = video.length() / fileSize.toDouble()
            setProgressValue(Tools.percentFormat.format(ratio))
        } else {
            setProgressValue("File Not Found")
        }
    }

    private fun setProgressValue(value: String) {
        if (progress.value != value) {
            progress.value = value
        }
    }

    fun update(download: Download) {
        fileSize = download.fileSize
        isDownloading = download.isDownloading
        updateProgress()
    }

    @get:JsonIgnore
    val downloadFile: File?
        get() {
            if (!StringChecker.isNullOrEmpty(downloadPath)) {
                val file = File(downloadPath)
                if (file.exists()) {
                    return file
                }
            }
            return null
        }

    @get:JsonIgnore
    val isComplete: Boolean
        get() {
            if (fileSize <= 0) {
                return false
            }
            val file = downloadFile
            return if (file != null) {
                if (file.length() <= 0) {
                    false
                } else file.length() >= fileSize
            } else false
        }
}